### DISCUSSION TITLE
Next 13 fetch function inside server component doesn't redirect routes

### DISCUSSION THREAD

**Discussion by obadakhalili** [vercel/next.js#43215] [Help]
I'm using Next 13 with an existing C# backend, and so I set up a custom server that redirects all `/api` calls to that backend and everything else to the next handler.
When I try to `fetch("/api")` inside a client component the call gets redirected correctly, but when I try to do the same inside a server component an error is thrown saying `unhandledRejection: TypeError: Failed to parse URL from /api`.

Is there a solution for this?

**Comment by icyJoseph** [comment]
Hi,

This is a bit of a catch-22 situation. Allow me to expand on this subject.

API routes require a server to run. They do not exist at build time for example. Now what's even more interesting is that servers have no origin attached to themselves, so a relative path is often not possible out of the box. As per standard, there's something called `basedURL` that fetch has to figure out and in Node.js contexts, it is not always possible AFAIK.

People then often do `new URL(path, baseURL)` and then it is up to you to figure out the `baseURL`. Doing this necessarily implies an extra jump for data, which is redundant because you are already on the server, so you could just connect to the service you need directly from there.

In GetServerSideProps `GSSP`, and of course GetStaticProps `GSP`, this was also an issue. GSSP required absolute URLs as well, it is nothing new.  Try it out, use `fetch("/api")` inside GSSP and it will have the same issue. Meanwhile `GSP` wouldn't work at all because the `api` routes simply did not exist at build time.

There's this piece of documentation from Next.js that people often disregard, https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes:

> It can be tempting to reach for an [API Route](https://nextjs.org/docs/api-routes/introduction) when you want to fetch data from the server, then call that API route from getServerSideProps. This is an unnecessary and inefficient approach, as it will cause an extra request to be made due to both getServerSideProps and API Routes running on the server.

The same rings true to Server Components, you can just reach for the 3rd party service you need directly from there, it is of no use to reach for an `api/route` from there. It is solvable via absolute URLs but it is not practical.

My favourite solution here is to refactor so that the contents of the handler in `api/route` can be reused anywhere, by parametrising the operation:

```ts
// parametrised to user, doesn't depend on HTTP request constructs
export const getCodeWarsUser = (user) =>
  axios
    .get(`https://www.codewars.com/api/v1/users/${user}`)
    .then(({ data }) => data);
```

And then:

```ts
import { getCodeWarsUser } from 'lib/codewars';

export async function getServerSideProps(context: GetServerSidePropsContext) {
  // calculate user from context
  const codewars = await getCodeWarsUser("some-user");

  return {
    props: { codewars },
  };
}
```

Or

```ts
// api/codewars.ts
import { getCodeWarsUser } from 'lib/codewars';

const codewars = async (req, res) => {
 // calculate "some-user" from req
  const data = await getCodeWarsUser('some-user');

  return res.send(data);
};

export default codewars;
```

Or with Server Components:

```tsx
export default async function Page({
  searchParams,
}: {
  searchParams: { user?: string };
}) {
  const { user } = searchParams;
  const data = await getCodeWarsUser(user);

  return (
    <section>
      <h2>Code wars users</h2>
    </section>
  );
}
```

**Reply by obadakhalili** [reply]
Thank you for the elaborate answer. Can I make from your answer that integrating with an existing backend in Next 13 isn't well-supported?

**Reply by icyJoseph** [reply]
Why would that be the conclusion? It is designed to integrate with backends transparently. What you are trying to do, it seems, is add an additional step to call upon your own API route, which is not necessary. 

Could you explain further:  `integrating with an existing backend in Next 13` what does existing backend mean here? I am quite shocked that you arrived to that conclusion, that's why I won't make any further assumptions, and let you explain your use case, and point of view further.

**Reply by icyJoseph** [reply]
Try with something like:

```tsx
import { redirect } from 'next/navigation';

const getRedirectLocation = async () => {
  // here localhost:3001/api/redirect is another project, it could also be a 3rd party, etc...
  const response = await fetch("http://localhost:3001/api/redirect", {
    redirect: "manual",
  });

  console.log(response.status, response.headers.get("location"));
  return response.headers.get("location");
};

export default async function Page() {
  // I guess here you could check page params or searchParams
  const location = await getRedirectLocation();
  if (location) {
   // do something with the location 
  redirect(location); 
 }
 // otherwise render some JSX or w.e.
  return (
    <section>
      <h1>hello world</h1>
    </section>
  );
}

```

All of the above is server code, so as you can see, it is pretty much the way to integrate with a backend... quite the opposite of your conclusion.

**Reply by obadakhalili** [reply]
Maybe I have misdelivered my point. I meant that integrating with a server **of a different language** that runs locally isn't well-supported.

**Reply by icyJoseph** [reply]
If it runs locally, as a separate process, then you can just call it by its full URL, no need to use relative paths.
